### DISCUSSION TITLE
Server with long running callback - send ACKNOWLEDGE to avoid timeout

### DISCUSSION THREAD

**Discussion by truschival** [pymodbus-dev/pymodbus#1788] [Q&A] [answered]
I am writing a server using ``RunAsyncTCPServer`` and following the example ``server_callback.py`` I created a  Datablock that overwrites ``getValues()`` and ``validate()``. 
My problem is that ``getValues()`` takes longer than my ModbusTCP client expects, I get timeout errors. According to the [Modbus Spec](https://www.modbus.org/docs/Modbus_Application_Protocol_V1_1b.pdf) page 49 I think I could send a  ``ACKNOWLEDGE`` message and later send the data using the same transaction id.  I have no clue how to send a message from the callback methods. 

I had a look at  ``request_tracer`` or ``response_manipulator`` but it seems they don't really help here - or at least I have no idea how.

How would you deal with the situation of long-running callbacks for reading data?

**Comment by janiversen** [comment, marked as answer]
You cannot send and Ack at that level !

You are expected to return an error if the data is not ready, then the client can retry later.

**Reply by truschival** [reply]
Thanks. I think I wrote the question before thinking it thought. 

I guess the better idea is to implement some sort of protocol on application layer i.e. 
1. write a request, 
2. poll for a finish-bit, 
3. read result
